miden-lib/asm/sat/internal/account_vault.masm

```diff
@@ -63,3 +66,227 @@ export.has_non_fungible_asset
+#       this ergonomically?
+
+#! Add the specified fungible asset to the vault.  If the vault already contains an asset
+#! issued by the same faucet, the amounts are added together.
+#!
+#! Panics:
+#! - If the total value of assets is greater than or equal to 2^63.
+#!
+#! Stack: [ASSET]
+#! Output: [ASSET']
+#!
+#! - ASSET is the fungible asset to add to the vault.
+#! - ASSET' is the updated fungible asset.
+proc.add_fungible_asset
+    push.0 movdn.3
+    # => [ASSET_KEY, amount]
+
+    # get the asset vault root and read asset
+    exec.layout::get_acct_vault_root swapw exec.smt::get
+    # => [CUR_ASSET, VAULT_ROOT, amount]
+
+    # arrange elements
+    movup.3 movup.8 dup
+    # => [amount, amount, cur_amount, faucet_id, 0, 0, VAULT_ROOT]
+
+    # compute max_amount - cur_amount
+    exec.asset::get_fungible_asset_max_amount dup.3 sub
+    # => [(max_amount - cur_amount), amount, amount, cur_amount, faucet_id, 0, 0, VAULT_ROOT]
+
+    # asset amount + cur_amount < max_amount
+    lte assert
+    # => [amount, cur_amount, faucet_id, 0, 0, VAULT_ROOT]
+
+    # add asset amounts
+    add movdn.3 
+    # => [ASSET', VAULT_ROOT]
+
+    # prepare the stack to insert the asset into the vault
+    dupw movdnw.2 dupw movup.3 drop push.0 movdn.3 swapw
+    # => [KEY, ASSET', VAULT_ROOT, ASSET']
+
+    # update asset in vault
+    exec.smt::insert dropw
+    # => [VAULT_ROOT', ASSET']
+
+    # update the vault root
+    exec.layout::set_acct_vault_root
+    # => [ASSET']
+end
+
+#! Add the specified non-fungible asset to the vault.
+#!
+#! Panics:
+#! - If the vault already contains the same non-fungible asset.
+#!
+#! Stack: [ASSET]
+#! Output: [ASSET]
+proc.add_non_fungible_asset
+    # prepare the stack to insert the asset into the vault
+    dupw exec.layout::get_acct_vault_root swapw dupw
+    # => [ASSET, ASSET, VAULT_ROOT, ASSET]
+
+    # insert asset into vault
+    exec.smt::insert
+    # => [OLD_VAL, VAULT_ROOT', ASSET]
+
+    # Assert old value was empty
+    padw assert_eqw
+    # => [VAULT_ROOT', ASSET]
+
+    # update the vault root
+    exec.layout::set_acct_vault_root
+    # => [ASSET]
+end
+
+#! Add the specified asset to the vault.
+#!
+#! Panics:
+#! - If the asset is not valid.
+#! - If the total value of two fungible assets is greater than or equal to 2^63.
+#! - If the vault already contains the same non-fungible asset.
+#!
+#! Stack: [ASSET]
+#! Output: [ASSET' / ASSET]
+#!
+#! - ASSET is the asset to add to the vault.
+#! - ASSET' is the updated fungible asset.
+export.add_asset
+    # assert that the asset is valid
+    exec.asset::validate_asset
+    # => [ASSET]
+
+    # check if the asset is a fungible asset
+    exec.asset::is_fungible_asset
+    # => [is_fungible_asset, ASSET]
+
+    # add the asset to the asset vault
+    if.true
+        exec.add_fungible_asset
+        # => [ASSET']
+    else
+        exec.add_non_fungible_asset
+        # => [ASSET]
+    end
+end
```

Comment:
We could do this a bit more efficiently as:
```
exec.asset::is_fungible_asset

if.true
    exec.asset::validate_fungible asset
    exec.add_fungible_asset
else
    exec.asset::validate_non_fungible_asset
    exec.add_non_fungible_asset
end
```
But I also wonder to which extend we need to validate the asset. If the user supplies an invalid asset, this should fail the asset preservation check, right?

Reply:
Yes you're right this would fail asset preservation checks so it's not technically required.  Maybe we follow a similar approach to your proposal for checking fungible asset amounts of leaving the check for now and then removing it later?  Happy to remove it now if you think it's preferable.

Reply:
Let's keep for now and create an issue to discuss this. I think it's worth thinking through this much more before removing the checks.